skglm/solvers/prox_newton.py

```diff
@@ -0,0 +1,338 @@
+        lipschitz[idx] = raw_hess @ X[:, j] ** 2
+
+    cached_grads = np.zeros(len(ws))
+    X_delta_w_ws = np.zeros(X.shape[0])
+    w_ws = w_epoch[ws]
+
+    for cd_iter in range(MAX_CD_ITER):
+        for idx, j in enumerate(ws):
+            # skip when X[:, j] == 0
+            if lipschitz[idx] == 0:
+                continue
+
+            cached_grads[idx] = grad_ws[idx] + X[:, j] @ (raw_hess * X_delta_w_ws)
+            old_w_idx = w_ws[idx]
+            stepsize = 1 / lipschitz[idx]
+
+            w_ws[idx] = penalty.prox_1d(
+                old_w_idx - stepsize * cached_grads[idx], stepsize, j)
+
+            if w_ws[idx] != old_w_idx:
+                X_delta_w_ws += (w_ws[idx] - old_w_idx) * X[:, j]
+
+        if cd_iter % 5 == 0:
+            opt = penalty.subdiff_distance(w_ws, cached_grads, ws)
+            if np.max(opt) <= tol:
+                break
+
+    # descent direction
+    return w_ws - w_epoch[ws], X_delta_w_ws
+
+
+# sparse version of _compute_descent_direction
+@njit
+def _descent_direction_s(X_data, X_indptr, X_indices, y, w_epoch,
+                         Xw_epoch, grad_ws, datafit, penalty, ws, tol):
+    raw_hess = datafit.raw_hessian(y, Xw_epoch)
+
+    lipschitz = np.zeros(len(ws))
+    for idx, j in enumerate(ws):
+        # equivalent to: lipschitz[idx] += raw_hess * X[:, j] ** 2
+        lipschitz[idx] = _sparse_squared_weighted_norm(X_data, X_indptr, X_indices,
+                                                       j, raw_hess)
+
+    cached_grads = np.zeros(len(ws))
+    X_delta_w_ws = np.zeros(len(y))
+    w_ws = w_epoch[ws]
+
+    for cd_iter in range(MAX_CD_ITER):
+        for idx, j in enumerate(ws):
+            # skip when X[:, j] == 0
+            if lipschitz[idx] == 0:
+                continue
+
+            cached_grads[idx] = grad_ws[idx]
+            # equivalent to cached_grads[idx] += X[:, j] @ (raw_hess * X_delta_w_ws)
+            cached_grads[idx] += _sparse_weighted_dot(X_data, X_indptr, X_indices, j,
+                                                      X_delta_w_ws, raw_hess)
+
+            old_w_idx = w_ws[idx]
+            stepsize = 1 / lipschitz[idx]
+
+            w_ws[idx] = penalty.prox_1d(
+                old_w_idx - stepsize * cached_grads[idx], stepsize, j)
+
+            if w_ws[idx] != old_w_idx:
+                _update_X_delta_w(X_data, X_indptr, X_indices, X_delta_w_ws,
+                                  w_ws[idx] - old_w_idx, j)
+
+        if cd_iter % 5 == 0:
+            opt = penalty.subdiff_distance(w_ws, cached_grads, ws)
+            if np.max(opt) <= tol:
+                break
+
+    # descent direction
+    return w_ws - w_epoch[ws], X_delta_w_ws
+
+
+@njit
+def _backtrack_line_search(X, y, w, Xw, datafit, penalty, delta_w_ws,
+                           X_delta_w_ws, ws):
+    # inplace update of w and Xw
+    # return grad_ws of the last w and Xw
+    step, prev_step = 1., 0.
+    epoch_penalty_val = penalty.value(w[ws])
```

Comment:
> `epoch_penalty_val`

I would change then name, maybe for `penalty_val_old`

Reply:
In addition, sadly you need to call 
`penalty.value(w)` because of the weighted Lasso penalty possibility. (Btw we might want to change this in another PR, and call penalty.value(w, ws))